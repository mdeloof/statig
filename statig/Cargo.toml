[package]
name = "statig"
version = "0.4.0"
edition = "2021"
rust-version = "1.66"
authors = ["Maxim Deloof"]
license = "MIT"
description = "Hierarchical state machines for designing event-driven systems"
readme = "../README.md"
repository = "https://github.com/mdeloof/statig"
keywords = ["fsm", "hsm", "statechart", "state-machine", "embedded"]

[dependencies]
statig_macro = { path = "../macro", version = "0.4.0", optional = true }
serde = { version = "1.0.152", optional = true }
bevy_ecs = { version = "0.12.1", optional = true }


[dev-dependencies]
bincode = "1.3.3"
futures = { version = "0.3.26" }
serde_json = "1.0.91"
serde = { version = "1.0.152", features = ["derive"] }
unit-enum = { git = "https://github.com/mdeloof/unit-enum.git" }

[features]
default = ["macro"]
macro = ["statig_macro"]
serde = ["dep:serde"]
bevy = ["dep:bevy_ecs"]
async = []
