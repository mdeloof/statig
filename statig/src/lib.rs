#![cfg_attr(not(doctest), doc = include_str!("../../README.md"))]
#![no_std]
#![allow(incomplete_features)]

mod response;
mod state_or_superstate;

/// Macro for deriving the state and superstate enum.
///
/// By parsing the underlying `impl` block and searching for methods with the
/// `state`, `superstate` or `action` attribute, the `state_machine` macro can
/// derive the state and superstate enums. By default these will be given the
/// names '`State`' and '`Superstate`'. Next to that the macro will also
/// implement the [`State`](crate::blocking::State) trait for the state enum and the
/// [`Superstate`](crate::blocking::Superstate) trait for the superstate enum.
///
/// To override the default configuration you can use the following attributes.
///
/// - `#[state_machine(state(name = "CustomStateName"))]`
///
///   Set the name of the state enum to a custom name.
///
///   _Default_: `State`
///
///   <br/>
///
/// - `#[state_machine(superstate(name = "CustomSuperstateName"))]`
///
///   Set the name of the superstate enum to a custom name.
///
///   _Default_: `Superstate`
///
///   <br/>
///
/// - `#[state_machine(state(derive(SomeTrait, AnotherTrait)))]`
///
///   Apply the derive macro with the passed traits to the state enum.
///
///   _Default_: `()`
///
///   <br/>
///
/// - `#[state_machine(superstate(derive(SomeTrait, AnotherTrait)))]`
///
///   Apply the derive macro with the passed traits to the superstate enum.
///
///   _Default_: `()`
///
///   <br/>
#[cfg(feature = "macro")]
pub use statig_macro::state_machine;

/// Attribute for tagging a state.
///
/// This macro does nothing on its own but is detected by the `state_machine`
/// macro when added to a method.
///
/// It accepts the following attributes:
///
/// - `#[state(name = "CustomStateName")]`
///
///   Set the name of the variant that will be part of the state enum.
///
///   <br/>
///
/// - `#[state(superstate = "superstate_name")]`
///
///   Set the superstate of the state.
///
///   <br/>
///
/// - `#[state(entry_action = "entry_action_name")]`
///
///   Set the entry action of the state.
///
///   <br/>
///
/// - `#[state(exit_action = "exit_action_name")]`
///
///   Set the exit action of the state.
///
///   <br/>
///
/// - `#[state(local_storage("field_name_a: FieldTypeA", "field_name_b: FieldTypeB"))]`
///
///   Add local storage to this state. These will be added as fields to the enum variant.
///
///   <br/>
#[cfg(feature = "macro")]
pub use statig_macro::state;

/// Attribute for tagging a superstate.
///
/// This macro does nothing on its own but is detected by the `state_machine`
/// macro when added to a method.
///
/// It accepts the following attributes:
///
/// - `#[superstate(name = "CustomSuperstateName")]`
///
///   Set the name of the variant that will be part of the state enum.
///
///   <br/>
///
/// - `#[superstate(superstate = "superstate_name")]`
///
///   Set the superstate of the superstate.
///
///   <br/>
///
/// - `#[superstate(entry_action = "entry_action_name")]`
///
///   Set the entry action of the superstate.
///
///   <br/>
///
/// - `#[superstate(exit_action = "exit_action_name")]`
///
///   Set the exit action of the superstate.
///
///   <br/>
///
/// - `#[superstate(local_storage("field_name_a: &'a mut FieldTypeA"))]`
///
///   Add local storage to this superstate. These will be added as fields to
///   the enum variant. It is crucial to understand that superstates never own
///   their data. Instead it is always borrowed from the underlying state or
///   superstate. This means the fields should be references with an
///   associated lifetime `'a`.
///
///   <br/>
#[cfg(feature = "macro")]
pub use statig_macro::superstate;

/// Attribute for tagging an action.
///
/// This macro does nothing on its own but is detected by the `state_machine`
/// macro when added to a method.
#[cfg(feature = "macro")]
pub use statig_macro::action;

/// Prelude containing the necessary imports for use with macro.
pub mod prelude {
    #![allow(ambiguous_glob_reexports)]
    #![allow(unused_imports)]

    #[cfg(any(feature = "async", doc))]
    pub use crate::awaitable::{IntoStateMachineExt as _, StateExt as _, *};
    pub use crate::blocking::{IntoStateMachineExt as _, StateExt as _, *};
    pub use crate::Response::{self, *};
    pub use crate::StateOrSuperstate;
    #[cfg(feature = "macro")]
    pub use statig_macro::state_machine;
}

pub mod blocking;

#[cfg(any(feature = "async", doc))]
pub mod awaitable;

pub use response::*;
pub use state_or_superstate::*;
