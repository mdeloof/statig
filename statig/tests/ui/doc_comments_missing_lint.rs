#![deny(missing_docs)]
//! Critical validation test: This MUST fail to prove our doc comment tests work.
//!
//! This test intentionally has missing doc comments on a state handler.
//! The #![deny(missing_docs)] lint should catch this and cause compilation to fail.
//!
//! If this test passes, it means either:
//! 1. Our doc comment implementation is broken, OR  
//! 2. The missing_docs lint isn't working
//!
//! Either case would invalidate our positive doc comment tests.

use statig::prelude::*;

/// Test machine for missing docs validation
#[derive(Default)]
pub struct ValidationMachine {}

/// State machine to validate missing docs detection
#[state_machine(initial = "State::documented()")]
impl ValidationMachine {
    /// This state has documentation.
    #[state]
    fn documented(&mut self) -> Outcome<State> {
        Transition(State::undocumented())
    }

    // INTENTIONALLY MISSING DOC COMMENT - this should cause compile failure
    #[state]
    fn undocumented(&mut self) -> Outcome<State> {
        Transition(State::documented())
    }

    /// This superstate has doc comments.
    #[superstate]
    fn container(&mut self) -> Outcome<State> {
        Super
    }
}

/// Main function for the test
fn main() {
    let machine = ValidationMachine::default();
    let _state_machine = machine.uninitialized_state_machine().init();
}
