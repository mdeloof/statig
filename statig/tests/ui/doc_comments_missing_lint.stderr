error: missing documentation for a variant
  --> tests/ui/doc_comments_missing_lint.rs:20:1
   |
20 | #[state_machine(initial = "State::documented()")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> tests/ui/doc_comments_missing_lint.rs:1:9
   |
 1 | #![deny(missing_docs)]
   |         ^^^^^^^^^^^^
   = note: this error originates in the attribute macro `state_machine` (in Nightly builds, run with -Z macro-backtrace for more info)
